pkg/workflow/workflow.go

```diff
@@ -0,0 +1,175 @@
+	}
+}
+
+func (w *workflow) ExecuteSteps(ctx context.Context, rev string, objects []*unstructured.Unstructured) (bool, error) {
+	steps := w.app.Spec.Workflow
+
+	if len(steps) == 0 {
+		return true, nil
+	}
+
+	w.app.Status.Phase = common.ApplicationRunningWorkflow
+
+	w.app.Status.Workflow = []common.WorkflowStepStatus{}
+	for i, step := range steps {
+		obj := objects[i].DeepCopy()
+		obj.SetName(step.Name)
+		obj.SetNamespace(w.app.Namespace)
+		obj.SetOwnerReferences([]metav1.OwnerReference{
+			*metav1.NewControllerRef(w.app, oamcore.ApplicationKindVersionKind),
+		})
+		err := w.applyWorkflowStep(ctx, obj, &types.WorkflowContext{
+			AppName:       w.app.Name,
+			AppRevision:   rev,
+			WorkflowIndex: i,
+			ResourceConfigMap: corev1.LocalObjectReference{
+				Name: rev,
+			},
+		})
+		if err != nil {
+			return false, err
+		}
+
+		status, err := w.syncWorkflowStatus(step, obj)
+		if err != nil {
+			return false, err
+		}
+
+		w.app.Status.Workflow = append(w.app.Status.Workflow, *status)
+		switch status.Phase {
+		case common.WorkflowStepPhaseSucceeded: // This one is done. Continue
+		case common.WorkflowStepPhaseRunning: // Need to retry shortly.
+			return false, nil
+		default:
+			return true, nil
+		}
+	}
+	return true, nil // all steps done
+}
+
+func (w *workflow) applyWorkflowStep(ctx context.Context, obj *unstructured.Unstructured, wctx *types.WorkflowContext) error {
+	if err := addWorkflowContextToAnnotation(obj, wctx); err != nil {
+		return err
+	}
+
+	return w.applicator.Apply(ctx, obj)
+}
+
+func addWorkflowContextToAnnotation(obj *unstructured.Unstructured, wc *types.WorkflowContext) error {
+	b, err := json.Marshal(wc)
+	if err != nil {
+		return err
+	}
+	m := map[string]string{
+		oam.AnnotationWorkflowContext: string(b),
+	}
+	obj.SetAnnotations(oamutil.MergeMapOverrideWithDst(m, obj.GetAnnotations()))
+	return nil
+}
+
+const (
+	// CondTypeWorkflowFinish is the type of the Condition indicating workflow progress
+	CondTypeWorkflowFinish = "workflow-progress"
+
+	// CondReasonSucceeded is the reason of the workflow progress condition which is succeeded
+	CondReasonSucceeded = "Succeeded"
+	// CondReasonStopped is the reason of the workflow progress condition which is stopped
+	CondReasonStopped = "Stopped"
+	// CondReasonFailed is the reason of the workflow progress condition which is failed
+	CondReasonFailed = "Failed"
+)
+
+func (w *workflow) syncWorkflowStatus(step oamcore.WorkflowStep, obj *unstructured.Unstructured) (*common.WorkflowStepStatus, error) {
+	status := &common.WorkflowStepStatus{
+		Name: step.Name,
+		Type: step.Type,
+		ResourceRef: runtimev1alpha1.TypedReference{
+			APIVersion: obj.GetAPIVersion(),
+			Kind:       obj.GetKind(),
+			Name:       obj.GetName(),
+			UID:        obj.GetUID(),
+		},
+	}
+
+	cond, found, err := utils.GetUnstructuredObjectStatusCondition(obj, CondTypeWorkflowFinish)
+	if err != nil {
+		return nil, err
+	}
+
+	if !found || cond.Status != "True" {
```

Comment:
Will you make `True` as a constant? And it could be a bool?

Reply:
I've check the status of a Pod, I doubt why there are two types of `true`?

```
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2021-05-26T13:20:10Z"
    reason: PodCompleted
    status: "True"
    type: Initialized
...
  initContainerStatuses:
  - containerID: docker://22e1a23c40335e7c811f869214122f1d1fa1d7bac1bd913ef0050e43bd0c3fa6
    image: busybox:latest
    imageID: docker-pullable://busybox@sha256:b5fc1d7b2e4ea86a06b0cf88de915a2c43a99a00b6b3c0af731e5f4c07ae8eff
    lastState: {}
    name: prepare-input-terraform-configurations
    ready: true
```

Reply:
Pod is not a very good example... The status conditions are standardized long after it.

Reply:
> Will you make True as a constant?

fixed